Adds completion for tab-focus.

toofar opened the pull request:
This is so you can jump between tabs based on tab title or URL instead
of just index just like the `buffer` command in vim. Adds a `b` alias
for `tab-focus` to be more like vim too.

The method of updating the completion model (rebuild the whole thing
every time a tab is created, removed or updated) is inefficient but is
still very fast for even unreasonable numbers of open tabs.

TODO: I did a quick test and it doesn't seem to work with multiple windows.
Probably because of how I am getting the `win_id`. Can anyone tell me how
to get the _current_ `win_id`?

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/the-compiler/qutebrowser/1317)
<!-- Reviewable:end -->

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -849,6 +850,11 @@ def tab_focus(self, index: {'type': (int, 'last')}=None, count=None):
+            try:
+                index = int(index)
+            except ValueError:
+                raise cmdexc.CommandError("tab-focus takes an int or 'last' not: {}".format(index))
```
What's the difference between checking this here and the way it was before (via the argument annotation, `index: {'type': (int, 'last')}=None`)?

(I didn't actually check, but they should do the exact same thing, no?)

toofar replied:
Oh I suppose it would do the same thing. Since typing until there is only one option left then pressing enter passes a string through I was going to add a title to index lookup there for one less keypress but that seems like way too much effort. I'll change it back.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,8 +832,9 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
     @cmdutils.register(instance='command-dispatcher', scope='window',
-                       count='count')
-    def tab_focus(self, index: {'type': (int, 'last')}=None, count=None):
+                       name=['tab-focus', 'b'], count='count',
```
How important is the `b` alias to you? vim compatibility for commands never really was a concern, and a `b` command is rather unintuitive, so I think I'd prefer users just setting their own aliases if they want to (though there's a long-standing issue with completions for aliases...)

toofar replied:
I suppose I am just used to it. Anyway I will add the buffer command and leave people to do their own thing like you say.

vyp replied:
FWIW I believe the default keybinding in pentadactyl puts this functionality under the `b` keybinding too, so it's probably not completely unintuitive, at least for those coming from pentadactyl. However, I also agree with leaving people to do their own thing.

The-Compiler commented:
Handling multiple windows obviously is the biggest problem here, yeah. The concept of a "current window" doesn't really exist when the completions are being rebuilt, and currently there's no mechanism for the completion to let the model know the current window when the completion is being *invoked*.

Those are good ideas to keep in mind for #74 somewhen in the future though.

I'd suggest a slightly different approach: Don't add a completion for `:tab-focus` for now, but add a different `:buffer` command to switch to *any* tab (in any window).

Then you can iterate over *all* windows (e.g. via `for win_id in objreg.window_registry:`) and build a completion for all of them, using a completion category for each window.

The-Compiler commented:
#32 is the related issue, by the way - a `:buffer` command is one of the most requested features and I never got to it, so thanks for tackling this, I'm sure many will appreciate it!

vyp commented:
> a `:buffer` command is one of the most requested features and I never got to it, so thanks for tackling this, I'm sure many will appreciate it!

aye aye! :sparkles: :smiley: :+1:

toofar commented:
Alrighty. I changed the completion model to iterate over all windows, put `tab-focus` back how it was, add a new `buffer` command which can switch focus between windows and can work based on a substring so that you can be lazy and don't have to actually select anything in the completion widget.

The-Compiler commented:
So I played with this a bit - seems quite nice, but it seems the completion doesn't update correctly when I do `:open -w example.com` (or open new tabs in the new window).

I'll also add some line comments for some style things - thanks for this again, looking forward to it!

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            index: The win_id/index of the tab to focus. Or a substring
+                   in which case the closest match will be focused.
+        """
+        if not index or not len(index):
```
Empty strings are false in Python, so checking `if not index:` here is enough.

On the other hand, why not just make `index` mandatory by removing the `=None` default?

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+
+        try:
+            # Is index of form 12/34 ?
+            list(map(int, index.split('/')))
```
I'd rather avoid using `map()` for its side-effect, I'd prefer this:

```python
for part in index.split('/'):
    int(part)
```

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+                tabbed_browser.setCurrentIndex(i-1)
+                return
+            except ValueError:
+                pass
```
I have a hard time understanding what all this exactly does, and the catched `ValueError` is way too broad, IMHO. Can you try refactoring it in a way it fails early whenever possible (i.e. having multiple `raise cmdexc.CommandError(...)` statements if necessary) and make the block where `ValueError` is ignored smaller?

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            if sf.count() > 0:
+                index = sf.data(sf.first_item())
+
+        if len(index.split('/')):
```
Isn't this always taken? Even `''.split('/')` will give you `['']`.

toofar replied:
Aha I think some of these are due to me having different ideas about how the
method would work, then changing strategies mid way through and not going back
and cleaning things up. Woopsie.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+                    aw = app.qApp.activeWindow()
+                    for win_id in objreg.window_registry:
+                        if aw == objreg.window_registry[win_id]:
+                            w = win_id
```
You should be able to simply do `active_win = objreg.get('app').activeWindow()` and `w = active_win.win_id` instead. Note you'll need to handle `active_win` being `None`.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,62 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+
+                tabbed_browser = objreg.get('tabbed-browser', scope='window',
+                                            window=w)
+                if i < 1 or i > tabbed_browser.count():
```
You can do `if not 0 < i <= tabbed_browser.count():` instead

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,67 @@ def __init__(self, parent=None):
+                              tab.url().toDisplayString(), 
+                              tabbed_browser.page_title(i))
+
+        if isinstance(arg, webview.WebView):
```
Why not simply `if arg is not None:`?

toofar replied:
Because that method is also called on `webview.url_text_changed` which emits a string.

The-Compiler replied:
This isn't obvious IMHO - can you split this into two functions instead, and connect them accordingly?

toofar commented:
but it seems the completion doesn't update correctly on new windows
Yeah ... coulda sworn that worked during testing. Looks like because there is no notification of new windows (or new `tabbed_browser`s) the completion model never gets a chance to hook up to the new `new_tab` signal.

toofar commented:
I cleaned up commands:buffer() and added a signal
`Application.new_window()` which is emited from the `MainWindow.__init__()`
because it looks like windows are made all over the place and the main
application is the only non-per-window object (?). Not entirely sure how elegent that is but it works better now!

The-Compiler commented:
I think that's the best solution - I'll leave some more line comments.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,51 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            idx = int(index)
+            win_id = 0
+            active_win = objreg.get('app').activeWindow()
+            if not active_win == None:
```
This should be `if active_win is not None:`

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,51 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            win_id = 0
+            active_win = objreg.get('app').activeWindow()
+            if not active_win == None:
+                win_id = active_win.win_id
```
I think you should handle the case where `active_win` is `None` explicitly - currently you just fall back to `win_id = 0` which means the first window, but even if you used `-1` or so the error ("There's no window with id -1!") would be confusing.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,51 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            raise cmdexc.CommandError(
+                "There's no tab with index {}!".format(idx))
+
+        objreg.get('app').setActiveWindow(objreg.window_registry[win_id])
```
The docs for `setActiveWindow` say:

> The function is called from the platform specific event handlers.

So I'm not sure if it's intended to be used like that.

I'd rather do something like this:

```python
window = objreg.window_registry[win_id]
window.activateWindow()
window.raise_()
```

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,71 @@ def __init__(self, parent=None):
+    def on_tab_close(self):
+        QTimer.singleShot(0, self.rebuild_cat)
+
+    def rebuild_cat(self, arg=None):
```
Please also decorate things connected to signals with `@pyqtSlot(<type>)` (you might need to import it from `PyQt5.QtCore`)

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,51 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+                    "No matching tab for: {}".format(index))
+
+        if len(index.split('/')) == 2:
+            win_id, idx = map(int, index.split('/'))
```
I don't like `map` :wink:

I think this would be better written as:

```python
index_parts = index.split('/')

if len(index_parts) == 2:
    win_id = int(index_parts[0])
    idx = int(index_parts[1])
```

(then you can also re-use `index_parts` for the `elif` below)

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,51 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+
+        if len(index.split('/')) == 2:
+            win_id, idx = map(int, index.split('/'))
+        elif len(index.split('/')) == 1:
```
What if none of those is true, e.g. the passed argument is `1/2/3`?

The-Compiler commented:
Can you please also try writing some end-to-end tests for this? You can find some inspiration in `tests/integration/features/tabs.feature`. You can run those locally by installing `tox` and optionally `Xvfb` (to avoid windows popping up), and then run `tox -e py35 -- tests/integration/features/test_tabs.py`. (or `py34` if you're on a system with Python 3.4)

toofar commented:
Sorry for taking so long getting back to you. Writing tests is not very motivating. Hope this addresses your concerns.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,60 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+                # Not sure how you enter a command without and active window...
+                raise cmdexc.CommandError(
+                    "No window specified and couldn't find active window!")
+            win_id = active_win.win_id
```
Can you add an `else: raise AssertionError("Unexpected index length {}".format(len(index_parts)))` or so here, so there's a defined behaviour if that should ever happen (e.g. by someone changing the completion model in the future) rather than undefined behaviour?

toofar replied:
Well there is a `index_parts = index.split('/', 1)` at the top of the method so there really isn't going to b more than 2 parts.

The-Compiler replied:
Fair enough, I missed that :smile:

The-Compiler commented on tests/integration/features/tabs.feature:
```diff
@@ -644,3 +644,86 @@ Feature: Tab management
+        When I open data/title.html
+        And I open data/search.html in a new tab
+        And I open data/scroll.html in a new tab
+        And I run :open -w http://localhost:(port)/data/caret.html
```
Can't you just do "And I open data/caret.html in a new window" here?

toofar replied:
eh, I suppose so. I was just copy/pasting.

The-Compiler commented:
Thank you! Tests are just necessary for a project of this scale (and it being cross-platform with different Qt versions), unfortunately I realized this much too late :wink:

There are still some small things to do, and then I'll merge this :smile:

- Fix the things mentioned in the line comments
- Add some tests for `:buffer` with invalid/strange arguments and make sure it does the right thing:
    - `:buffer -1`
    - `:buffer /`
    - `:buffer //`
    - `:buffer 0/x`
    - `:buffer 1/2/3`
- Fix the style issues flagged by vulture/pylint/flake:

```
qutebrowser/completion/models/miscmodels.py:154: Unused variable 'IDX_COLUMN'
```

```
./qutebrowser/browser/commands.py:837:1: D205 1 blank line required between summary line and description
./qutebrowser/browser/commands.py:837:1: D400 First line should end with a period
./qutebrowser/completion/models/instances.py:61:1: E302 expected 2 blank lines, found 1
./qutebrowser/completion/models/miscmodels.py:145:1: E302 expected 2 blank lines, found 1
./qutebrowser/completion/models/miscmodels.py:170:45: E127 continuation line over-indented for visual indent
./qutebrowser/completion/models/miscmodels.py:190:1: D202 No blank lines allowed after function docstring
./qutebrowser/completion/models/miscmodels.py:204:45: E127 continuation line over-indented for visual indent
./qutebrowser/completion/models/miscmodels.py:210:59: W291 trailing whitespace
```

```
************* Module qutebrowser.mainwindow.mainwindow
W: 27, 0: Unused pyqtSignal imported from PyQt5.QtCore (unused-import)
************* Module qutebrowser.completion.models.miscmodels
C:210, 0: Trailing whitespace (trailing-whitespace)
W:205,30: Access to a protected member _shutting_down of a client class (protected-access)
```

toofar commented:
Regarding the checks for strange arguments I am not sure what is the right behaviour for some of them. Everything after the first `/` in the argument is taken to be a part of a tab title or url which is supposed to be freeform. If you pass something like a slash in it matches the first url it compares to which is correct given the input even if a little simplistic. The only way around that while stil leaving the fuzzy matching thing in is to add special cases. Also for duplicates it currently selects the first match (by tab index based on the current sorting algorithm), vim for comparison just gives a `More than one match for {}` error.

The-Compiler commented:
I'm okay with the current behaviour, I just want some tests to make sure there are no crashes with such arguments.

toofar commented:
Okay, finally got around to doing that stuff that shouldn't have taken this long. FYI when I ran `tox -e py35 -- tests/integration/features/test_tabs.py` I got

    ...
    py35 runtests: commands[1] | ~/projects/qutebrowser/.tox/py35/bin/python -m py.test tests/integration/features/test_tabs.py
    ~/projects/qutebrowser/.tox/py35/bin/python: loader for pytest cannot handle py.test
    ...

google led me to https://github.com/benoitc/gunicorn/issues/440 and running the failing command replacing `py.test` with `pytest` works. I'm on debian testing and just upgraded to python3.5, this didn't happen with 3.4 *shrugs*.

The-Compiler commented:
Thank you, and no worries!

I've seen that one recently too, I guess I'll do the same in `tox.ini`.

I'll give the PR another review round later this week, currently a bit busy too.

The-Compiler commented:
I'll add some more line comments - if you're currently short on time, just let me know, and I'll fix those myself instead :smile:

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -832,6 +833,61 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
+            idx = int(index_parts[1])
+        elif len(index_parts) == 1:
+            idx = int(index_parts[0])
+            win_id = -1
```
No point in doing this when you're setting `win_id` in line 873 (and bailing out otherwise) anyways, right?

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,80 @@ def __init__(self, parent=None):
+        few/few dozen/few hundred tabs doesn't take very long at all.
+        """
+        # working with a tabbed_browser whilst shutting down throws some errors
+        # pylint: disable=protected-access
```
I think the proper thing to do here would be to make `tabbed_browser._shutting_down` public (by removing the `_` everywhere it's used in `qutebrowser/mainwindow/tabbedbrowser.py`) instead.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,80 @@ def __init__(self, parent=None):
+        make sure we handled background loads too ... but iterating over a
+        few/few dozen/few hundred tabs doesn't take very long at all.
+        """
+        # working with a tabbed_browser whilst shutting down throws some errors
```
How can I reproduce this? Would it be possible to write a test for it, or is this already tested by simply starting/quitting qutebrowser?

toofar replied:
Okay I just removed that check and it still gets called when closing down
(from `on_tab_close()`) but no exceptions are thrown anymore. From memory it was
complaining about `tabbed_browser.widget(i)` or something because some of the Qt
stuff had been freed already. 

As to reproducing it, it wasn't a fatal error it was just stack traces being
printed but everything else worked fine. But I can't reproduce it now anyway.
And it is tested by exiting after initialising tab completion.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,80 @@ def __init__(self, parent=None):
+                              tab.url().toDisplayString(),
+                              tabbed_browser.page_title(i))
+
+        if isinstance(arg, webview.WebView):
```
As mentioned before, I'd really like this to be two separate method instead - this is really confusing IMHO

Link-Satonaka commented:
fwiw, I've been @toofar 's tab-complete branch with no issues. If I try to access `:buffers` before the browser finishes starting up, the list will be empty, but otherwise all is good.

The-Compiler commented:
Thanks for letting me know! Now that the crowdfunding is up I can get back to coding and reviewing PRs, so I'll probably merge this in the coming few days.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,78 @@ def __init__(self, parent=None):
+        QTimer.singleShot(0, self.rebuild_cat)
+
+    @pyqtSlot(str)
+    def rebuild_cat(self, arg=None): # pylint: disable=unused-argument
```
What's the reason for the argument here? It seems to work fine when I use `@pyqtSlot()` and drop the `, arg=None`. Note slots can have less arguments than the signals they're connected to.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -138,3 +141,78 @@ def __init__(self, parent=None):
+        make sure we handled background loads too ... but iterating over a
+        few/few dozen/few hundred tabs doesn't take very long at all.
+        """
+        self.removeRows(0, len(self._cats))
```
Couldn't you just use `self.rowCount()` here and not store `self._cats` at all?

The-Compiler commented:
When I close a window, I still see the window category when typing `:buffer` until I close the next tab - I think that's just due to `objreg.window_registry` still containing the windows for a little bit until everything is cleaned up? Not sure if you can do much about this, though...

Could you also fix the line comments and the few nitpicks found by the linters?

```
************* Module qutebrowser.completion.models.miscmodels
C:187, 4: Missing method docstring (missing-docstring)
```

```
./qutebrowser/completion/models/miscmodels.py:197:37: E261 at least two spaces before inline comment
```

Can you also run `scripts/dev/src2asciidoc.py` and commit the result?

After that, I think this is ready to merge - thanks a lot!

toofar commented:
> When I close a window, I still see the window category when typing :buffer until I close the next tab - I think that's just due to objreg.window_registry still containing the windows for a little bit until everything is cleaned up? Not sure if you can do much about this, though...

Since I already added a tabbed_browser.new_window signal I suppose I could add a window_close signal too.

The-Compiler commented:
Thanks for the quick reaction!

> Since I already added a tabbed_browser.new_window signal I suppose I could add a window_close signal too.

Sounds good - can you try if that helps? I'm not entirely sure (due to how `window_registry` works), but I think it should.

What do you think about keybindings for `:set-cmd-text -s :buffer`? Since we normally conform to dwb keybindings, that'd be `gt`, but that's already bound to `:tab-focus` (by @thiagowfx in #443)... I'm somewhat inclined to just change that to conform with dwb like everything else.

toofar commented:
Turns out the model was getting rebuilt after a window closed (due to the `tab_closed` signal) but the categories were being created before `tabbed_browser.shutting_down` was checked. So that was an easy fix. (Conecting to the `MainWindow.destroyed` signal and setting a timer also worked.)

For the keybindings I am just using `b`. I'm not really a fan of most of the dwb keybindings, I would prefer them to more ... logical; to have more of grammer than almost arbitrary mnemonics.

The-Compiler commented:
Merged - thanks for the great work, and sorry for the delays!

The-Compiler commented:
It's been a couple of years - but since a long time, we can get the current `win_id` for the completion, so I [added](https://github.com/qutebrowser/qutebrowser/commit/305e7c96d5e2fdb3b248b27dfb21042fb2b7e0b8) `:tab-focus` completion based on the `:buffer` completion now!

rien333 commented:
Thanks for still being committed to PRs as old as this one! I understood that the PR backlog was a bit of hurdle, so good luck with the other ones you plan to do!
